pkg/service/repair/worker.go

```diff
@@ -8,271 +8,158 @@ import (
+	defer func() {
+		w.logger.Info(ctx, "Repair done")
+		// Try to justify error by checking table deletion
+		if out != nil && w.isTableDeleted(ctx, j) {
+			out = errTableDeleted
 		}
+		out = errors.Wrapf(out, "master %s keyspace %s table %s command %d", j.master, j.keyspace, j.table, jobID)
+	}()
 
-		return errors.Wrapf(err, "host %s: schedule repair", host)
+	jobID, err = w.client.Repair(ctx, j.keyspace, j.table, j.master, j.replicaSet, j.tryOptimizeRanges())
+	if err != nil {
+		return errors.Wrap(err, "schedule repair")
 	}
 
-	w.progress.OnScyllaJobStart(ctx, job, jobID)
-	defer w.progress.OnScyllaJobEnd(ctx, job, jobID)
-
-	logger := w.logger.With(
-		"keyspace", ttr.Keyspace,
-		"table", ttr.Table,
-		"hosts", ttr.Replicas,
-		"ranges", len(ttrs),
-		"master", host,
+	w.logger.Info(ctx, "Repairing",
+		"keyspace", j.keyspace,
+		"table", j.table,
+		"master", j.master,
+		"hosts", j.replicaSet,
+		"ranges", j.tryOptimizeRanges(),
 		"job_id", jobID,
 	)
 
-	logger.Info(ctx, "Repairing")
-	if err := w.waitRepairStatus(ctx, jobID, host, ttr.Keyspace, ttr.Table); err != nil {
-		return errors.Wrapf(err, "host %s: keyspace %s table %s command %d", host, ttr.Keyspace, ttr.Table, jobID)
-	}
-	logger.Info(ctx, "Repair done")
-
-	return nil
-}
-
-func (w *worker) rowLevelRepair(ctx context.Context, job job) error {
-	err := w.runRepair(ctx, job)
+	status, err := w.client.RepairStatus(ctx, j.master, jobID)
 	if err != nil {
-		w.logger.Error(ctx, "Run row-level repair", "error", err)
-	}
-	return err
-}
-
-func (w *worker) legacyRepair(ctx context.Context, job job) error {
-	p := w.hostPartitioner[job.Host]
-
-	// In a very rare case when there is a strange partitioner
-	// do not split to shards.
-	if p == nil {
-		err := w.runRepair(ctx, job)
-		if err != nil {
-			w.logger.Error(ctx, "Run legacy repair, no sharding due to unsupported partitioner", "error", err)
-		}
-		return err
-	}
-
-	// Calculate max parallel shard repairs
-	limit := int(p.ShardCount())
-	if job.Allowance.ShardsPercent != 0 {
-		l := float64(limit) * job.Allowance.ShardsPercent
-		if l < 1 {
-			l = 1
-		}
-		limit = int(l)
-
-		w.logger.Debug(ctx, "Limiting parallel shard repairs", "total", p.ShardCount(), "limit", limit)
+		return errors.Wrap(err, "get repair status")
 	}
 
-	// Split ranges to shards
-	shardRanges, err := splitToShardsAndValidate(job.Ranges, p)
-	if err != nil {
-		return errors.Wrap(err, "split to shards")
-	}
-
-	f := func(i int) error {
-		if ctx.Err() != nil {
-			return nil
-		}
-
-		if len(shardRanges[i]) == 0 {
-			return nil
-		}
-
-		shardJob := job
-		shardJob.Ranges = shardRanges[i]
-		return w.runRepair(log.WithFields(ctx, "subranges_of_shard", i), shardJob)
-	}
-
-	notify := func(i int, err error) {
-		w.logger.Error(ctx, "Run legacy repair", "error", err)
+	switch status {
+	case scyllaclient.CommandRunning:
+		return w.handleRunningStatus(ctx, j)
+	case scyllaclient.CommandFailed:
+		return errors.Errorf("status %s", status)
+	default:
```

Comment:
It's risky.
Any response other than "RUNNING", "SUCCESFULL", "FAILED" should be considered as an error.

nil error is interpreted by the manager as "SUCCESFULL" -> https://github.com/scylladb/scylla-manager/blob/beb0ae93c055c9674b4c38b1b663944e5ef4970b/pkg/service/repair/generator.go#L89-L92